### PR TITLE
Fix reading bad channels

### PR DESCRIPTION
**Pull request by withmywoessner**
#### Reference issue
Fixes: [11832](https://github.com/mne-tools/mne-python/issues/11832)

#### What does this implement/fix?
Fixes bug where `read_raw_cnt()` does not read bad channels

### CODE REVIEW

**Comment by withmywoessner** [comment]
Actually, I think I need to do some more investigating. The sample data set already includes a test to check for bad channels and it uses byte 5 instead of 15 and it passes. However, the documentation cites it is located at byte 15. http://paulbourke.net/dataformats/eeg/sethead.h

**Comment by larsoner** [comment]
Looks like this is where we look for bad channel names

https://github.com/mne-tools/mne-python/blob/74f3f18c82d3548e12833712ac6fd6074f1a3005/mne/io/cnt/cnt.py#L345-L351

**Comment by larsoner** [comment]
> Is @jaeilepp still around?

No and this was implemented a long time ago IIRC so I doubt there would be much memory remaining. Maybe older vs newer versions of the file spec or something?

**Comment by withmywoessner** [comment]
I thought so too, but the strange thing was that a comment written at that
time cites documentation with the new file format.

**Comment by larsoner** [comment]
We find bugs in our code all the time :) If the documentation *and* the results on files where you know what to expect in terms of good/bad channels say we're doing something wrong currently, then we probably are

**Comment by withmywoessner** [comment]
> > My change will fail reading that sample data. Should I remove it?
> 
> Have you checked to see if that file has any version information in it? That way we could triage which bad-channel-reading method to use. Otherwise we could use a try/except but that's a bit uglier / brittle.

It does not have any version info. Another discrepancy is that the file has 128 channels, but when I try reading it into Neuroscan it states that it does not support 128 channels.

**Comment by larsoner** [comment]
Probably we have an "old" format file that Neuroscan no longer supports then. It would be good to be able to support both. Maybe try/except is the best we can do... or we add a `header='old' | 'new'` kwarg or something

**Comment by withmywoessner** [comment]
> Probably we have an "old" format file that Neuroscan no longer supports then. It would be good to be able to support both. Maybe try/except is the best we can do... or we add a `header='old' | 'new'` kwarg or something

Okay, I can do that! 

I just find it strange that the documentation references a website that describes the 'new' file format. If it's based on the old format why does it reference new documentation? Maybe the website changed?

**Comment by larsoner** [comment]
For some reason a bunch of the `tools/*.sh` got `chmod -x`'ed

**Comment by withmywoessner** [comment]
> For some reason a bunch of the `tools/*.sh` got `chmod -x`'ed

Yes, sorry about that @larsoner I am using wsl2, and apparently [there is a problem with configuring git line endings](https://code.visualstudio.com/docs/remote/troubleshooting#_resolving-git-line-ending-issues-in-containers-resulting-in-many-modified-files) that causes this. I reverted the push. I ended up adding a keyword, does the documentation on the website get updated automatically?

File: mne/io/cnt/cnt.py [hunk 1]
```diff
@@ -173,6 +173,7 @@ def read_raw_cnt(
     date_format="mm/dd/yy",
     preload=False,
     verbose=None,
+    header="new",
```
Comment (larsoner) [inline mne/io/cnt/cnt.py:176]: Should default to `"old"` for backward compatibility

File: mne/io/cnt/tests/test_cnt.py [hunk 2]
```diff
@@ -19,11 +19,15 @@
...
+def test_old_data():
     """Test reading raw cnt files."""
     with pytest.warns(RuntimeWarning, match="number of bytes"):
         raw = _test_raw_reader(
-            read_raw_cnt, input_fname=fname, eog="auto", misc=["NA1", "LEFT_EAR"]
+            read_raw_cnt,
+            input_fname=fname,
+            eog="auto",
+            misc=["NA1", "LEFT_EAR"],
+            header="old",
```
Comment (larsoner) [inline mne/io/cnt/tests/test_cnt.py:30]: Then this can be reverted

File: mne/io/cnt/tests/test_cnt.py [hunk 3]
```diff
@@ -37,11 +41,20 @@ def test_data():
...
+
+    assert raw.info["bads"] == ["F8"]  # test bads
+
+
 @testing.requires_testing_data
 def test_compare_events_and_annotations():
     """Test comparing annotations and events."""
     with pytest.warns(RuntimeWarning, match="Could not parse meas date"):
-        raw = read_raw_cnt(fname)
+        raw = read_raw_cnt(fname_bad_spans)
```
Comment (larsoner) [inline mne/io/cnt/tests/test_cnt.py:57]: Shouldn't need to change this, right?

**Review by larsoner — COMMENTED** [review]
Failures look related, e.g.:

https://github.com/mne-tools/mne-python/actions/runs/6369190963/job/17289232256?pr=11833

Can you reproduce locally?

File: mne/io/cnt/tests/test_cnt.py (hunk 3 above)
Comment (withmywoessner) [reply mne/io/cnt/tests/test_cnt.py:57]: Yes, that doesn't need to be changed. I was going to revert it but I am having trouble with the line endings.

**Comment by withmywoessner** [comment]
Okay, everything should be fixed now @larsoner. I think the only thing that needs to be done is the testing data needs to be updated. I already merged the new file in: 
https://github.com/mne-tools/mne-testing-data/blob/master/CNT/test_CNT_events_mne_JWoess_clipped.cnt

**Comment by larsoner** [comment]
That PR has been merged and in MNE-Python `main` it should exist, but the CIs are not happy:
```
________________________________ test_new_data _________________________________
mne/io/cnt/tests/test_cnt.py:46: in test_new_data
    assert raw.info["bads"] == ["F8"]  # test bads
E   AssertionError: assert [] == ['F8']
E     Right contains one more item: 'F8'
E     Full diff:
E     - ['F8']
E     + []
```

**Comment by withmywoessner** [comment]
Okay! Thanks! @larsoner. I found the problem and fixed it.

**Comment by larsoner** [comment]
Does the test pass locally for you? Remotely it doesn't:

https://github.com/mne-tools/mne-python/actions/runs/6385705361/job/17331050081?pr=11833#step:17:4172

Can you try:
```
pytest mne/io/cnt/tests/test_cnt.py -k new_data
```
? It's best to iterate locally like this first before pushing to make sure things are green

**Comment by withmywoessner** [comment]
Yes @larsoner, I just ran the test and it passed. I will be sure to be more mindful about looking over my code if I make another contribution. Thanks for the help!

File: mne/io/cnt/cnt.py [hunk 4]
```diff
@@ -171,6 +171,7 @@ def read_raw_cnt(
     emg=(),
     data_format="auto",
     date_format="mm/dd/yy",
+    header="old",
```
Comment (larsoner) [inline mne/io/cnt/cnt.py:174]: To make sure nobody accidentally passses `preload` by position accidentally to `header`
```suggestion
    *,
    header="old",
```

File: mne/io/cnt/cnt.py [hunk 5]
```diff
@@ -480,6 +493,7 @@ def __init__(
         emg=(),
         data_format="auto",
         date_format="mm/dd/yy",
+        header="old",
```
Comment (larsoner) [inline mne/io/cnt/cnt.py:496]:
```suggestion
        *,
        header="old",
```

**Comment by larsoner** [comment]
Okay last thing, can you add a line to `doc/changes/devel.rst`?

**Pull request closed without merging**